### PR TITLE
Improve the impl to allow  the explicit initialisation in the annotation

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3682

## Checklist
- [x] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 1]
```diff
@@ -73,11 +73,14 @@ public enum DiagnosticsCodes {
...
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
-    PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_124("PERSIST_124", "the entity ''{0}'' should be a closed record.", ERROR),
     PERSIST_125("PERSIST_125", "relation annotation cannot be attached without the entity annotation",
             ERROR),
-    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without the " +
-            "entity annotation", ERROR),
+    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without the entity " +
+            "annotation", ERROR),
```
Comment (daneshk) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:80]:
```suggestion
    PERSIST_126("PERSIST_126", "auto increment annotation can only be attached to the entity", ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 2]
```diff
@@ -73,11 +73,14 @@ public enum DiagnosticsCodes {
             "references for relation", ERROR),
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
-    PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_124("PERSIST_124", "the entity ''{0}'' should be a closed record.", ERROR),
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:76]: Lets follow the same convention in all codes. Use a simple case to start.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java (hunk 1 above)
Comment (niveathika) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:80]:
```suggestion
    PERSIST_126("PERSIST_126", "auto increment annotation can only be attached to an entity, ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 3]
```diff
@@ -73,11 +73,14 @@ public enum DiagnosticsCodes {
...
     PERSIST_125("PERSIST_125", "relation annotation cannot be attached without the entity annotation",
             ERROR),
-    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without the " +
-            "entity annotation", ERROR),
+    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without the entity " +
+            "annotation", ERROR),
+    PERSIST_127("PERSIST_127", "invalid value initialization: a field only allows explicit " +
+            "initialization",
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:82]:
```suggestion
    PERSIST_127("PERSIST_127", "this field only allows inline initialisation",
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 4]
```diff
@@ -73,11 +73,14 @@ public enum DiagnosticsCodes {
             "references for relation", ERROR),
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
-    PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_124("PERSIST_124", "the entity ''{0}'' should be a closed record.", ERROR),
     PERSIST_125("PERSIST_125", "relation annotation cannot be attached without the entity annotation",
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]: PERSIST_125 and PERSIST_117 is identically the same.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 5]
```diff
@@ -308,14 +305,22 @@ private void validateEntityAnnotation(SyntaxNodeAnalysisContext ctx, TypeDefinit
                                 if (!name.equals(Constants.TABLE_NAME)) {
                                     validateEntityProperties(ctx, expressionNode.get(), tableFields);
                                 } else {
-                                    tableName = Utils.eliminateDoubleQuotes(expressionNode.get().toSourceCode().trim());
-                                    if (tableNames.contains(tableName)) {
-                                        reportDiagnosticInfo(ctx, mappingFieldNode.location(),
+                                    ExpressionNode exp = expressionNode.get();
+                                    if (exp instanceof BasicLiteralNode) {
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:309]: This validation needs to be done for the array itself as well as individual elements

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 6]
```diff
@@ -308,14 +305,22 @@ private void validateEntityAnnotation(SyntaxNodeAnalysisContext ctx, TypeDefinit
...
+                                        if (tableNames.contains(tableName)) {
+                                            reportDiagnosticInfo(ctx, mappingFieldNode.location(),
+                                                    DiagnosticsCodes.PERSIST_113.getCode(),
+                                                    DiagnosticsCodes.PERSIST_113.getMessage(),
+                                                    DiagnosticsCodes.PERSIST_113.getSeverity());
+                                        } else {
+                                            tableNames.add(tableName);
+                                        }
+                                    } else {
+                                        reportDiagnosticInfo(ctx, exp.location(),
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:320]: Are we continuing processing even if this validation is shown?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java (hunk 2 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:76]: Checked.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java (hunk 4 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java]: Updated in e4a2137

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java (hunk 5 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:309]: Already added this validation array as well https://github.com/ballerina-platform/module-ballerina-persist/pull/73/files#diff-2fef4b0b023d4af527c920976d479032ef537155e9422e1cfd555c006d0a6d96R374

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java (hunk 6 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:320]: Yes

**Review by niveathika — APPROVED** [review]

**Pull request merged**